src/common/dout.h

```diff
@@ -43,14 +43,48 @@ class DoutPrefixProvider {
   virtual ~DoutPrefixProvider() {}
 };
 
+// helpers
+namespace ceph {
+namespace dout {
+
+template<typename T>
+struct dynamic_marker_t {
+  T value;
+  operator T() const { return value; }
+};
+
+template<typename T>
+dynamic_marker_t<T> need_dynamic(T&& t) {
+  return dynamic_marker_t<T>{ std::forward<T>(t) };
+}
+
+template<typename T>
+struct is_dynamic : public std::false_type {};
+
+template<typename T>
+struct is_dynamic<dynamic_marker_t<T>> : public std::true_type {};
+
+} // dout
+} // ceph
+
 // generic macros
 #define dout_prefix *_dout
 
 #define dout_impl(cct, sub, v)						\
   do {									\
-  if (cct->_conf->subsys.should_gather(sub, v)) {			\
+  const bool should_gather = [&](const auto cctX) {			\
+    if constexpr (ceph::dout::is_dynamic<decltype(sub)>::value) {	\
+      return cctX->_conf->subsys.should_gather(sub, v);			\
+    } else {								\
+      return cctX->_conf->subsys.template should_gather<sub>(v);	\
+    }									\
+  }(cct);								\
+									\
+  if (unlikely(should_gather)) {					\
```

Comment:
Maybe we could differentiate the hinting policy basing on e.g. `v` and the default gathering level. Though, I'm not sure it's really necessary.